From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rezoss <rezoss@unica>
Date: Fri, 24 Jul 2026 10:30:00 +0200
Subject: [PATCH] Add Telegram class text parser fallback

Telegram clients expose chat rows through custom views with empty resource
ids, so the normal idEntry-based parser exits the chat session before
SmartReplyHandler receives messages. Add a Telegram-only fallback that builds
minimal ConversationMessage records from captured visible text when the normal
parser cannot produce a title or filtered message list.
---
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/LiveTranslationParser.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/LiveTranslationParser.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/LiveTranslationParser.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/LiveTranslationParser.smali
@@ -1486,6 +1486,15 @@
 
     if-nez v1, :cond_a
 
+    invoke-static {p3, v5}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->fallbackParse(Ljava/lang/String;Ljava/util/Map;)Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ParserResult;
+
+    move-result-object p0
+
+    if-eqz p0, :cond_rezoss_no_title_fallback
+
+    return-object p0
+
+    :cond_rezoss_no_title_fallback
     const-string/jumbo p0, "title is null"
 
     invoke-static {p1, p0}, Landroid/util/Log;->d(Ljava/lang/String;Ljava/lang/String;)I
@@ -1561,6 +1570,21 @@
 
     move-result-object p0
 
+    invoke-interface {p0}, Ljava/util/List;->isEmpty()Z
+
+    move-result p2
+
+    if-eqz p2, :cond_rezoss_messages_ok
+
+    invoke-static {p3, v5}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->fallbackParse(Ljava/lang/String;Ljava/util/Map;)Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ParserResult;
+
+    move-result-object p2
+
+    if-eqz p2, :cond_rezoss_messages_ok
+
+    return-object p2
+
+    :cond_rezoss_messages_ok
     invoke-interface {p0}, Ljava/util/List;->size()I
 
     move-result p2
diff --git a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge.smali b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge.smali
--- a/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge.smali
+++ b/smali_classes15/com/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge.smali
@@ -351,6 +351,302 @@
     return p0
 .end method
 
+.method public static final isTelegramPackage(Ljava/lang/String;)Z
+    .locals 2
+
+    if-eqz p0, :cond_false
+
+    const-string/jumbo v0, "org.telegram.messenger"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string/jumbo v0, "org.telegram.messenger.web"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string/jumbo v0, "org.telegram.plus"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string/jumbo v0, "org.thunderdog.challegram"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_false
+
+    :cond_true
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_false
+    const/4 p0, 0x0
+
+    return p0
+.end method
+
+.method private static final isTelegramNoiseText(Ljava/lang/String;)Z
+    .locals 2
+
+    if-eqz p0, :cond_true
+
+    invoke-static {p0}, Lkotlin/text/v;->isBlank(Ljava/lang/CharSequence;)Z
+
+    move-result v0
+
+    if-nez v0, :cond_true
+
+    const-string v0, "Message"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string/jumbo v0, "Send"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string v0, "Attach"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string v0, "Emoji"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string v0, "Search"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result v1
+
+    if-nez v1, :cond_true
+
+    const-string v0, "Back"
+
+    invoke-static {p0, v0}, Lbl0/c;->h(Ljava/lang/Object;Ljava/lang/Object;)Z
+
+    move-result p0
+
+    if-eqz p0, :cond_false
+
+    :cond_true
+    const/4 p0, 0x1
+
+    return p0
+
+    :cond_false
+    const/4 p0, 0x0
+
+    return p0
+.end method
+
+.method public static final fallbackParse(Ljava/lang/String;Ljava/util/Map;)Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ParserResult;
+    .locals 15
+    .annotation system Ldalvik/annotation/Signature;
+        value = {
+            "(",
+            "Ljava/lang/String;",
+            "Ljava/util/Map<",
+            "Landroid/view/autofill/AutofillId;",
+            "+",
+            "Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationView;",
+            ">;)",
+            "Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ParserResult;"
+        }
+    .end annotation
+
+    invoke-static {p0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isTelegramPackage(Ljava/lang/String;)Z
+
+    move-result v0
+
+    if-eqz v0, :cond_null
+
+    if-eqz p1, :cond_null
+
+    move-object/from16 v14, p1
+
+    new-instance v0, Ljava/util/ArrayList;
+
+    invoke-direct {v0}, Ljava/util/ArrayList;-><init>()V
+
+    new-instance v1, Ljava/util/HashSet;
+
+    invoke-direct {v1}, Ljava/util/HashSet;-><init>()V
+
+    const/4 v2, 0x0
+
+    invoke-interface {v14}, Ljava/util/Map;->values()Ljava/util/Collection;
+
+    move-result-object v3
+
+    check-cast v3, Ljava/lang/Iterable;
+
+    invoke-interface {v3}, Ljava/lang/Iterable;->iterator()Ljava/util/Iterator;
+
+    move-result-object v3
+
+    :goto_loop
+    invoke-interface {v3}, Ljava/util/Iterator;->hasNext()Z
+
+    move-result v4
+
+    if-eqz v4, :cond_after_loop
+
+    invoke-interface {v3}, Ljava/util/Iterator;->next()Ljava/lang/Object;
+
+    move-result-object v4
+
+    check-cast v4, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationView;
+
+    if-eqz v4, :goto_loop
+
+    invoke-virtual {v4}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationView;->getText()Ljava/lang/CharSequence;
+
+    move-result-object v5
+
+    if-eqz v5, :goto_loop
+
+    invoke-static {v5}, Lkotlin/text/v;->isBlank(Ljava/lang/CharSequence;)Z
+
+    move-result v6
+
+    if-eqz v6, :cond_has_text
+
+    goto :goto_loop
+
+    :cond_has_text
+    invoke-virtual {v5}, Ljava/lang/Object;->toString()Ljava/lang/String;
+
+    move-result-object v5
+
+    invoke-static {v5}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->isTelegramNoiseText(Ljava/lang/String;)Z
+
+    move-result v6
+
+    if-eqz v6, :cond_not_noise
+
+    goto :goto_loop
+
+    :cond_not_noise
+    if-nez v2, :cond_keep_title
+
+    move-object v2, v4
+
+    :cond_keep_title
+    invoke-interface {v1, v5}, Ljava/util/Set;->add(Ljava/lang/Object;)Z
+
+    move-result v6
+
+    if-eqz v6, :goto_loop
+
+    invoke-interface {v0}, Ljava/util/List;->size()I
+
+    move-result v6
+
+    const/16 v7, 0x1e
+
+    if-lt v6, v7, :cond_add_message
+
+    goto :goto_loop
+
+    :cond_add_message
+    new-instance v6, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationMessage;
+
+    invoke-virtual {v4}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationView;->getTop()I
+
+    move-result v10
+
+    invoke-virtual {v4}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationView;->getHeight()I
+
+    move-result v11
+
+    add-int/2addr v11, v10
+
+    invoke-virtual {v4}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationView;->getAFId()Landroid/view/autofill/AutofillId;
+
+    move-result-object v12
+
+    move-object v7, p0
+
+    move-object v8, v5
+
+    const/4 v9, 0x0
+
+    const/4 v13, 0x0
+
+    const/4 v14, 0x0
+
+    invoke-direct/range {v6 .. v14}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationMessage;-><init>(Ljava/lang/String;Ljava/lang/String;ZIILandroid/view/autofill/AutofillId;ZZ)V
+
+    invoke-interface {v0, v6}, Ljava/util/List;->add(Ljava/lang/Object;)Z
+
+    goto :goto_loop
+
+    :cond_after_loop
+    invoke-interface {v0}, Ljava/util/List;->isEmpty()Z
+
+    move-result p0
+
+    if-eqz p0, :cond_has_messages
+
+    goto :cond_null
+
+    :cond_has_messages
+    if-eqz v2, :cond_null
+
+    const-string p0, "Telegram"
+
+    invoke-virtual {v2, p0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationView;->setText(Ljava/lang/CharSequence;)V
+
+    const-string p0, "LiveTranslationParser"
+
+    const-string v1, "telegram class/text fallback parsed conversation"
+
+    invoke-static {p0, v1}, Landroid/util/Log;->i(Ljava/lang/String;Ljava/lang/String;)I
+
+    new-instance p0, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ParserResult;
+
+    const/4 v1, 0x0
+
+    invoke-direct {p0, v2, v0, v1}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ParserResult;-><init>(Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/ConversationView;Ljava/util/List;Landroid/graphics/Rect;)V
+
+    return-object p0
+
+    :cond_null
+    const/4 p0, 0x0
+
+    return-object p0
+.end method
+
 .method public static final populateAppViewFallback(Ljava/lang/String;Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/AppViewParsingSpec;)Z
     .locals 3
 
@@ -548,6 +844,22 @@
 
     invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
 
+    const-string v0, "org.telegram.ui.Components.RecyclerListView"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v0, "org.telegram.ui.ChatActivity$18"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v0, "org.telegram.ui.ChatActivity$ChatActivityFragmentView"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
+    const-string v0, "org.telegram.ui.Cells.ChatMessageCell"
+
+    invoke-static {p0, v0}, Lcom/samsung/android/smartsuggestions/service/deepscreencapture/screenunderstanding/NowNudgeConversationBridge;->add(Ljava/util/Set;Ljava/lang/String;)V
+
     const-string p0, "ScreenContentManager"
 
     const-string/jumbo v0, "Now Nudge conversation app-view fallback enabled"
